### ISSUE TITLE
Controlling when normalization requests are processed

### ISSUE THREAD

**Issue by nikswamy** [FStarLang/FStar#1674]
We encode normalization hints using Pervasives.norm and its variants (e.g., normalize_term etc.)

When the normalizer encounters a `norm steps e`, it reduces `e` with the reduction strategy described by `steps`.

However, the normalizer may not encounter this term when the user expects. There are several such situations

1. The programmer writes

```let f x1..xn = ... norm steps e ... ```

expecting that the body of the definition of `f` will be replaced with its normalization. However, definitions are never normalized on their own, except prior to extraction. So, this `norm steps e` remains in the definition of `f` during typechecking.

2. The programmer writes

```let f x1..xn : norm steps t = e```

When checking the type of `e` against the programmer-provided annotation, the annotated type `norm steps t` is normalized. This may or may not be a good thing: if the normalization is blocked on reducing one of the free variables `x` , then the reduced term will be a partially reduced mess. In other situations, it may be essential to normalize this type in order to actually check `e` against it.

3. F* may infer a term that contains a norm in it and a subsequent phase may then go ahead an normalize it. E.g, 

```
let as_normal_t (#a:Type) (x:a) : normal a = x

let f  = as_normal g
```

Here F* infers the type of `f` as `normal t` for some type `t`. But, then a later phase (e.g., in phase 2 of the typechecker, or during --use_extracted_interfaces), this inferred type will be reduced, forcing the typechecker + Z3 to check the equality of two potentially large, partially reduced terms.

4. In cases where `normalize_term` isn't reduced prior to SMT encoding, we're left with an uninterpreted function `normalize_term a` in the SMT context, and proofs usually fail at this point, unless you call  

```
let normalize_term_spec (#a: Type) (x: a) : Lemma (normalize_term #a x == x) = ()
```

There are probably other glitches like these that I'm forgetting.

Based on a chat with @Chris-Hawblitzel  today, we came up with the following proposal:

1. Add a new variant of `norm`, called, say, `norm_if_guard` 

```
noeq type term : Type u#(max a b) = 
 | T : #a:Type u#a -> a -> term 
 
abstract let norm_if_guard (guard:list term) (s:steps) (#a:_) (x:a) = x
```

The idea is that `norm_if_guard [ T t1; ...;  T tn] steps e` reduces `e` if and only if all the terms `t1...tn` themselves reduce to terms whose head-symbol is a data constructor or a constant (e.g., an integer, string literal etc.)

This gives the programmer some control over when to actually attempt to reduce `e`, avoiding reducing it in cases where it will just get stuck.

2. Then, prior to introducing any term into the context (whether it is a definition in TcEnv, a term encoded to SMT etc.) we should always make a pass over the term processing any norm requests in the term.

---

Ideally, we would eventually deprecate the current flavors of `norm` in favor of variants of `norm_if_guard`.

This is adding quite a bit of complexity and I'm not sure it's really solving all of the four problems I mentioned to start with, e.g., even with this change, it may be the case that a earlier phases of the typechecker produce `norm_if_guard` terms that subsequent phases normalize away, perhaps requiring large proofs of type equality.

But, the existing flavors of `norm` are hard to control and predict, so we need to do something.

Requesting feedback from @aseemr @mtzguido et al.
